In [51]:
import pandas as pd

### Reading and exploring the dataset

In [2]:
chicago = pd.read_csv('pandas/chicago.csv')
chicago.head(3)

Name    Position Title   Department Employee Annual Salary
0    AARON,  ELVIA J  WATER RATE TAKER  WATER MGMNT              $90744.00
1  AARON,  JEFFERY M    POLICE OFFICER       POLICE              $84450.00
2     AARON,  KARINA    POLICE OFFICER       POLICE              $84450.00

- All the text data is in uppercase :-/
- `Name` column has a comma between last name and the first name
- Currency is not formatted as it should be

In [3]:
chicago.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32063 entries, 0 to 32062
Data columns (total 4 columns):
Name                      32062 non-null object
Position Title            32062 non-null object
Department                32062 non-null object
Employee Annual Salary    32062 non-null object
dtypes: object(4)
memory usage: 501.0+ KB


**Trying to optimize the DataFrame.**

Counting the number of unique values in the `Department` column

In [4]:
chicago['Department'].nunique()

35

In [5]:
chicago['Department'].count()

32062

**We see that there are only 35 unique values in the `Department` column out of a total of 32062 values**

It is advisable to convert this column to category to save memory usage.

In [6]:
chicago['Department'] = chicago['Department'].astype('category')
chicago.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32063 entries, 0 to 32062
Data columns (total 4 columns):
Name                      32062 non-null object
Position Title            32062 non-null object
Department                32062 non-null category
Employee Annual Salary    32062 non-null object
dtypes: category(1), object(3)
memory usage: 408.0+ KB


# Common string methods `.lower( )`, `.upper( )`, `.title( )`, `.len( )`

In [7]:
chicago = pd.read_csv('pandas/chicago.csv')
chicago['Department'] = chicago['Department'].astype('category')
chicago.head(3)

Name    Position Title   Department Employee Annual Salary
0    AARON,  ELVIA J  WATER RATE TAKER  WATER MGMNT              $90744.00
1  AARON,  JEFFERY M    POLICE OFFICER       POLICE              $84450.00
2     AARON,  KARINA    POLICE OFFICER       POLICE              $84450.00

## Using these methods with regular Python strings

In [8]:
'Hello WORLD'.upper()

'HELLO WORLD'

In [9]:
'Hello WORLD'.lower()

'hello world'

In [10]:
'Hello WORLD'.title()

'Hello World'

In [11]:
len('Hello World')

11

**Running these methods on a Pandas Series result in an error**

The correct way is to use the `.str.` as demonstrated below

In [20]:
result = chicago['Department'].str.lower()
result.head(3)

0    water mgmnt
1         police
2         police
Name: Department, dtype: object

In [21]:
result = chicago['Name'].str.lower().str.title()
result.head(3)

0      Aaron,  Elvia J
1    Aaron,  Jeffery M
2       Aaron,  Karina
Name: Name, dtype: object

**`len( )` function on a Python string returns the length of the string.**

However when called on a Series of strings, the `len( )` returns the total number of values in the Series and **not** the length of each string in the series

In [22]:
len(chicago['Name'])

32063

**To obtain the length of each string in the Series use the `.str.len( )` method instead of the built-in Python `len( )` function**

In [24]:
chicago['Name'].str.len()

0        15.0
1        17.0
2        14.0
3        19.0
4        19.0
5        15.0
6        17.0
7        17.0
8        20.0
9        21.0
10       21.0
11       16.0
12       16.0
13       17.0
14       22.0
15       15.0
16       21.0
17       19.0
18       18.0
19       19.0
20       24.0
21       19.0
22       16.0
23       20.0
24       19.0
25       20.0
26       20.0
27       22.0
28       17.0
29       23.0
         ... 
32033    15.0
32034    14.0
32035    15.0
32036    15.0
32037    16.0
32038    11.0
32039    14.0
32040    16.0
32041    18.0
32042    18.0
32043    16.0
32044    17.0
32045    15.0
32046    14.0
32047    15.0
32048    15.0
32049    22.0
32050    18.0
32051    18.0
32052    16.0
32053    19.0
32054    14.0
32055    13.0
32056    16.0
32057    19.0
32058    18.0
32059    17.0
32060    19.0
32061    19.0
32062     NaN
Name: Name, Length: 32063, dtype: float64

# The string `.replace( )` method
### In a regular Python string

In [25]:
'Hello World'.replace('l', 'm')

'Hemmo Wormd'

### In a Pandas Series

In [28]:
chicago = pd.read_csv('pandas/chicago.csv').dropna(how='all')
chicago['Department'] = chicago['Department'].astype('category')
chicago.tail(2)

Name           Position Title Department  \
32060  ZYRKOWSKI,  CARLO E           POLICE OFFICER     POLICE   
32061  ZYSKOWSKI,  DARIUSZ  CHIEF DATA BASE ANALYST       DoIT   

      Employee Annual Salary  
32060              $87384.00  
32061             $113664.00

In [30]:
chicago['Department'].head(5)

0         WATER MGMNT
1              POLICE
2              POLICE
3    GENERAL SERVICES
4         WATER MGMNT
Name: Department, dtype: category
Categories (35, object): [ADMIN HEARNG, ANIMAL CONTRL, AVIATION, BOARD OF ELECTION, ..., STREETS & SAN, TRANSPORTN, TREASURER, WATER MGMNT]

**Replacing `MGMNT` with `MANAGEMENT`**

In [32]:
chicago['Department'] = chicago['Department'].str.replace('MGMNT', 'MANAGEMENT')
chicago.head(5)

Name            Position Title        Department  \
0      AARON,  ELVIA J          WATER RATE TAKER  WATER MANAGEMENT   
1    AARON,  JEFFERY M            POLICE OFFICER            POLICE   
2       AARON,  KARINA            POLICE OFFICER            POLICE   
3  AARON,  KIMBERLEI R  CHIEF CONTRACT EXPEDITER  GENERAL SERVICES   
4  ABAD JR,  VICENTE M         CIVIL ENGINEER IV  WATER MANAGEMENT   

  Employee Annual Salary  
0              $90744.00  
1              $84450.00  
2              $84450.00  
3              $89880.00  
4             $106836.00

**Working with the `Employee Annual Salary` column**

In [34]:
chicago['Employee Annual Salary'] = chicago['Employee Annual Salary'].str.replace('$','').astype('float')

In [35]:
chicago['Employee Annual Salary'].head(3)

0    90744.0
1    84450.0
2    84450.0
Name: Employee Annual Salary, dtype: float64

**Since the values in the `Employee Annual Salary` column are now `float`, we can perform operations such as sum, average on them.**

In [36]:
chicago['Employee Annual Salary'].sum()

2571506375.36

In [37]:
chicago['Employee Annual Salary'].mean()

80204.178633899

In [38]:
chicago['Employee Annual Salary'].std()

25098.329867510587

In [39]:
chicago['Employee Annual Salary'].median()

84450.0

In [40]:
chicago['Employee Annual Salary'].nlargest(n=10) # 10 largest values

8184     300000.0
7954     216210.0
25532    202728.0
8924     197736.0
8042     197724.0
19208    195000.0
3706     187680.0
18556    187680.0
29466    187680.0
13754    185364.0
Name: Employee Annual Salary, dtype: float64

---
# Filtering with string methods

In [41]:
chicago = pd.read_csv('pandas/chicago.csv').dropna(how='all')
chicago['Department'] = chicago['Department'].astype('category')
chicago.tail(2)

Name           Position Title Department  \
32060  ZYRKOWSKI,  CARLO E           POLICE OFFICER     POLICE   
32061  ZYSKOWSKI,  DARIUSZ  CHIEF DATA BASE ANALYST       DoIT   

      Employee Annual Salary  
32060              $87384.00  
32061             $113664.00

**Extracting only to those rows that contain the word 'WATER' in the `Position Title` column**

In [44]:
mask = chicago['Position Title'].str.lower().str.contains('water')
mask.head(3)

0     True
1    False
2    False
Name: Position Title, dtype: bool

In [45]:
chicago[mask]

Name                                 Position Title  \
0              AARON,  ELVIA J                               WATER RATE TAKER   
554         ALUISE,  VINCENT G             FOREMAN OF WATER PIPE CONSTRUCTION   
671            ANDER,  PERRY A                               WATER CHEMIST II   
685        ANDERSON,  ANDREW J  DISTRICT SUPERINTENDENT OF WATER DISTRIBUTION   
702          ANDERSON,  DONALD             FOREMAN OF WATER PIPE CONSTRUCTION   
1054          ASHLEY,  KARMA T                               WATER CHEMIST II   
1079         ATKINS,  JOANNA M                               WATER CHEMIST II   
1181        AZEEM,  MOHAMMED A                               WATER CHEMIST II   
1285            BAJIC,  JOHN A                          WATER METER MACHINIST   
2400          BOLTON,  BRIAN E                               WATER RATE TAKER   
2586           BOYCE,  ADNER L                               WATER CHEMIST II   
2745          BRANDYS,  DANIEL                               WATER CHEMIST II   
3143          BROWN,  SHARON L                               WATER RATE TAKER   
3246      BUCHANAN,  ANTHONY L                               WATER CHEMIST II   
3456             BURT,  CARL S                               WATER RATE TAKER   
3626            CAGE,  OPHELIA                               WATER RATE TAKER   
3663     CALDERONE,  MICHAEL P             FOREMAN OF WATER PIPE CONSTRUCTION   
4889            CLARK,  MARK A                          WATER METER MACHINIST   
5506         COOPER JR,  EDDIE                               WATER CHEMIST II   
6337      DAVENPORT,  CLARENCE                          WATER METER MACHINIST   
6421       DAVIS III,  WALLACE                   GEN SUPT OF WATER MANAGEMENT   
6564      DECALUWE,  STANLEY J             FOREMAN OF WATER PIPE CONSTRUCTION   
6954            DIAZ,  OSCAR A                               WATER RATE TAKER   
7356         DOWDY,  TIMOTHY J             FOREMAN OF WATER PIPE CONSTRUCTION   
7463         DRIVER,  CLAUDE A             FOREMAN OF WATER PIPE CONSTRUCTION   
7641       DURANT,  PATRICIA B                               WATER RATE TAKER   
7650        DURHAM,  PATRICK L                          WATER METER MACHINIST   
8092        ESPINOSA,  RODOLFO                               WATER RATE TAKER   
8660   FITZPATRICK,  CHARLES W             FOREMAN OF WATER PIPE CONSTRUCTION   
9325       GAITHER,  RASHEDA K                            CHIEF WATER CHEMIST   
...                        ...                                            ...   
23187      QUINTANILLA,  EDWIN                               WATER CHEMIST II   
24003         RIOS,  FRANCISCO                               WATER RATE TAKER   
24178         ROBERTS,  AKILAH                               WATER CHEMIST II   
24266         ROBINSON,  JERRY                           WATER METER ASSESSOR   
24422     RODRIGUEZ,  EDWARD L                           WATER METER ASSESSOR   
24501      RODRIGUEZ,  MARCO A                               WATER RATE TAKER   
24945       RUDMIN,  KENNETH D                    WATER QUALITY INSPECTOR I/C   
25055       RUSSNAK,  THOMAS W                           WATER METER ASSESSOR   
25233           SALEH,  HASSEN         DIR OF WATER PURIFICATION LABORATORIES   
25260      SALINAS,  EDUARDO S                 ENGINEER OF WATER PURIFICATION   
26372   SIDDIQUI,  ABDURRAZZAQ                              WATER CHEMIST III   
26519         SIMS,  DEMETRIUS                               WATER RATE TAKER   
26625           SKOWRON,  JOHN                               WATER RATE TAKER   
26871          SMITH,  NANCY A                               WATER RATE TAKER   
27015        SOJKA,  JEFFREY A                           WATER METER ASSESSOR   
27368      STANG,  CHRISTOPHER             FOREMAN OF WATER PIPE CONSTRUCTION   
27537    STEVENSON JR,  WALTER             FOREMAN OF WATER PIPE CONSTRUCTION   
27635       STOKES,  CAROLYN M                           

**Extracting only those rows that have 'Water' (case insensitive) only at the beginning in the `Position Title` column**

In [48]:
mask = chicago['Position Title'].str.lower().str.startswith('water')
mask.head(3)

0     True
1    False
2    False
Name: Position Title, dtype: bool

In [49]:
chicago[mask]

Name               Position Title   Department  \
0              AARON,  ELVIA J             WATER RATE TAKER  WATER MGMNT   
671            ANDER,  PERRY A             WATER CHEMIST II  WATER MGMNT   
1054          ASHLEY,  KARMA T             WATER CHEMIST II  WATER MGMNT   
1079         ATKINS,  JOANNA M             WATER CHEMIST II  WATER MGMNT   
1181        AZEEM,  MOHAMMED A             WATER CHEMIST II  WATER MGMNT   
1285            BAJIC,  JOHN A        WATER METER MACHINIST  WATER MGMNT   
2400          BOLTON,  BRIAN E             WATER RATE TAKER  WATER MGMNT   
2586           BOYCE,  ADNER L             WATER CHEMIST II  WATER MGMNT   
2745          BRANDYS,  DANIEL             WATER CHEMIST II  WATER MGMNT   
3143          BROWN,  SHARON L             WATER RATE TAKER  WATER MGMNT   
3246      BUCHANAN,  ANTHONY L             WATER CHEMIST II  WATER MGMNT   
3456             BURT,  CARL S             WATER RATE TAKER  WATER MGMNT   
3626            CAGE,  OPHELIA             WATER RATE TAKER  WATER MGMNT   
4889            CLARK,  MARK A        WATER METER MACHINIST  WATER MGMNT   
5506         COOPER JR,  EDDIE             WATER CHEMIST II  WATER MGMNT   
6337      DAVENPORT,  CLARENCE        WATER METER MACHINIST  WATER MGMNT   
6954            DIAZ,  OSCAR A             WATER RATE TAKER  WATER MGMNT   
7641       DURANT,  PATRICIA B             WATER RATE TAKER  WATER MGMNT   
7650        DURHAM,  PATRICK L        WATER METER MACHINIST  WATER MGMNT   
8092        ESPINOSA,  RODOLFO             WATER RATE TAKER  WATER MGMNT   
10170        GODINEZ,  JULIO A        WATER METER MACHINIST  WATER MGMNT   
10457        GORDON,  DONALD R        WATER METER MACHINIST  WATER MGMNT   
10707           GREEN,  JOHN F        WATER METER MACHINIST  WATER MGMNT   
11857          HEDRICK,  LEE A             WATER RATE TAKER  WATER MGMNT   
12408      HJELMGREN,  JAMES A        WATER METER MACHINIST  WATER MGMNT   
12487              HO,  KWOK L  WATER CONSERVATION ENGINEER  WATER MGMNT   
12594               HO,  NAM C             WATER CHEMIST II  WATER MGMNT   
12781     HOWELL JR,  GEORGE B             WATER CHEMIST II  WATER MGMNT   
13361           JACOB,  LOVELY            WATER CHEMIST III  WATER MGMNT   
13476        JANSKY,  THOMAS A        WATER METER MACHINIST  WATER MGMNT   
...                        ...                          ...          ...   
20936  O CONNOR JR,  MICHAEL J        WATER METER MACHINIST  WATER MGMNT   
20962          ODIBO,  GRACE E             WATER CHEMIST II  WATER MGMNT   
21078     OLBERA SR,  MARTIN R        WATER METER MACHINIST  WATER MGMNT   
21212         O NEAL,  ROZELLA         WATER METER ASSESSOR  WATER MGMNT   
21788           PARKER,  TERRY             WATER CHEMIST II  WATER MGMNT   
22384    PFANNKUCHE,  JOSEPH W        WATER METER MACHINIST  WATER MGMNT   
22916         PRICE,  JOHNNY L        WATER METER MACHINIST  WATER MGMNT   
23102          PUTZ,  ANDREA R        WATER QUALITY MANAGER  WATER MGMNT   
23187      QUINTANILLA,  EDWIN             WATER CHEMIST II  WATER MGMNT   
24003         RIOS,  FRANCISCO             WATER RATE TAKER  WATER MGMNT   
24178         ROBERTS,  AKILAH             WATER CHEMIST II  WATER MGMNT   
24266         ROBINSON,  JERRY         WATER METER ASSESSOR  WATER MGMNT   
24422     RODRIGUEZ,  EDWARD L         WATER METER ASSESSOR  WATER MGMNT   
24501      RODRIGUEZ,  MARCO A             WATER RATE TAKER  WATER MGMNT   
24945       RUDMIN,  KENNETH D  WATER QUALITY INSPECTOR I/C  WATER MGMNT   
25055       RUSSNAK,  THOMAS W         WATER METER ASSESSOR  WATER MGMNT   
26372   SIDDIQUI,  ABDURRAZZAQ            WATER CHEMIST III  WATER MGMNT   
26519         SIMS,  DEMETRIUS             WATER RATE TAKER  WATER MGMNT   
26625           SKOWRON,  JOHN             WATER RATE TAKER  WATER MGMNT   
26871          SMITH,  NANCY A             WATER RATE TAKER  WATER MGMNT   
27015        SOJKA,  JEFFREY A         WATER METER ASSESSOR  WATER MGMNT  

**Extracting the rows in which the value of the `Position Title` column ends with a `ist` (case insensitive)**

In [50]:
mask = chicago['Position Title'].str.lower().str.endswith('ist')
chicago[mask]

Name                          Position Title  \
184             AFROZ,  NAYYAR                            PSYCHIATRIST   
308           ALARCON,  LUIS J              LOAN PROCESSING SPECIALIST   
422           ALLAIN,  CAROLYN    SENIOR TELECOMMUNICATIONS SPECIALIST   
472             ALLEN,  ROBERT                               MACHINIST   
705        ANDERSON,  EDWARD M               SR PROCUREMENT SPECIALIST   
1022            ARTEAGA,  PAUL                               MACHINIST   
1163           AYALA JR,  JUAN             FIELD SANITATION SPECIALIST   
1285            BAJIC,  JOHN A                   WATER METER MACHINIST   
1558       BARRETT,  BARBARA J           TECHNICAL TRAINING SPECIALIST   
1869        BELTRAN,  MAURICIO                  PROCUREMENT SPECIALIST   
2095        BIGGANE,  THOMAS M                     SECURITY SPECIALIST   
2319            BOBBA,  MAHITA              GRANTS RESEARCH SPECIALIST   
2519      BOSTON,  NATHANIEL K             FIELD SANITATION SPECIALIST   
2541     BOUKYDIS,  NIKOLETA N                            PSYCHOLOGIST   
2634         BOYLE,  RICHARD A                               MACHINIST   
2678           BRADLEY,  JEENA              GRANTS RESEARCH SPECIALIST   
3975          CARLSEN,  JOHN T                            PSYCHOLOGIST   
4109            CARTER,  ALICE                       SAFETY SPECIALIST   
4479             CHAN,  JOSEPH               SR PROCUREMENT SPECIALIST   
4820             CIVITO,  JOHN                               MACHINIST   
4851         CLARKE,  DANIEL M                               MACHINIST   
4889            CLARK,  MARK A                   WATER METER MACHINIST   
5043         COCO,  MICHELLE T                 INVESTIGATOR SPECIALIST   
5049          CODY,  MICHAEL J                               MACHINIST   
5307          CONDON,  KRISTIN                   HR RECORDS SPECIALIST   
5341           CONNELLY,  JOHN                               MACHINIST   
5535          COPP,  VINCENT F                            PSYCHOLOGIST   
5633        CORTES,  JEZIEEL T               SR PROCUREMENT SPECIALIST   
5921         CRUMP,  CAROLYN R             FIELD SANITATION SPECIALIST   
5938             CRUZ,  ELBA T                            PSYCHOLOGIST   
...                        ...                                     ...   
26982         SNOW,  MICHAEL J             FIELD SANITATION SPECIALIST   
27209        SPEAR,  RICHARD D                               MACHINIST   
27306         SRIVER,  NIKKI K        PR TELECOMMUNICATIONS SPECIALIST   
27576        STEWART JR,  JOHN               SR PROCUREMENT SPECIALIST   
27593        STEWART,  STACY P               SR PROCUREMENT SPECIALIST   
27628       STOCKOVER,  THOMAS                               MACHINIST   
27755      STROUD JR,  JAMES E                   WATER METER MACHINIST   
27835        SULKES,  SHEILA A                            PSYCHOLOGIST   
27910         SURACE,  FRANK V                               MACHINIST   
28077        SZYMUSIAK,  DIANE                  PROCUREMENT SPECIALIST   
28353       TERRELL,  LYNNETTE               SR PROCUREMENT SPECIALIST   
28477       THOMAS,  MICHAEL E  REHABILITATION CONSTRUCTION SPECIALIST   
28520      THOMPSON,  CLAUDE O                               MACHINIST   
28578       THROWER,  DERWIN K                       SAFETY SPECIALIST   
29277            VACCARO,  JOE                               MACHINIST   
29310          VALDEZ,  JAVIER                               MACHINIST   
29502     VAUGHN JR,  JOSEPH T  REHABILITATION CONSTRUCTION SPECIALIST   
30060            WALSH,  MORAG              SENIOR ARCHIVAL SPECIALIST   
30156           WARD,  SONJI S               SR PROCUREMENT SPECIALIST   
30287         WATKINS,  GARY G           SR LABOR RELATIONS SPECIALIST   
30914        WILLIAMS,  JANA Y                  PROCUREMENT SPECIALIST   
31118          WILLIS,  KIRK J          CRIMES SURVEILLANCE SPECIALIST   
31296     WISNIEWSKI,  JERRY V

---
# `.strip( )`, `.lstrip( )`, and `.rstrip( )` methods
These methods are used to remove whitespaces from the strings

In [52]:
chicago = pd.read_csv('pandas/chicago.csv').dropna(how='all')
chicago['Department'] = chicago['Department'].astype('category')
chicago.head(3)

Name    Position Title   Department Employee Annual Salary
0    AARON,  ELVIA J  WATER RATE TAKER  WATER MGMNT              $90744.00
1  AARON,  JEFFERY M    POLICE OFFICER       POLICE              $84450.00
2     AARON,  KARINA    POLICE OFFICER       POLICE              $84450.00

### Using these methods with regular Python strings

In [53]:
'     Hello World    '.strip()

'Hello World'

In [54]:
'     Hello World    '.lstrip()

'Hello World    '

In [55]:
'     Hello World    '.rstrip()

'     Hello World'

### Using these methods with Pandas Series

In [58]:
chicago['Name'] = chicago['Name'].str.lstrip().str.rstrip()

In [60]:
chicago['Position Title'] = chicago['Position Title'].str.strip()